src/lib/encounters.lua

```diff
@@ -35,36 +40,44 @@ def generate_encounter(target_cr: int, seed: int):
     else:
         cr_min = max(target_cr - 3, 0)
         cr_max = target_cr
+        
+    cr_min = max(ceil(cr_min), 0) # the subtracted number only changes the range of monsters at cr 1, where it can also include monsters of >= cr 0.125
 
     monsters = monster_tools.get_monsters_in_cr_range(cr_min, cr_max)
 
-    # if there's no monsters in that cr range search lower
-    while len(monsters) == 0 and cr_min >= 0:
-        cr_min = floor(cr_min - 1)
+    # if there's no monsters in that cr range, expand range
+    while len(monsters) == 0 and cr_min > 0:
+        cr_min = cr_min - 1
         monsters = monster_tools.get_monsters_in_cr_range(cr_min, cr_max)
 
-    cr_min = max(cr_min, 0)
+    # if there is still no monsters, throw exception
+    if len(monsters) == 0:
+        {}[f"No monsters found in cr range {cr_min} to {cr_max}, please check if your server config has any monsters in this range."]
 
-    monsters_len = len(monsters)
     cr_budget = target_cr
     final_list = []
 
     if difficulty > 3:
-        highest_cr_mon = monsters[0]
+        highest_cr = max(monsters.keys())
+        highest_cr_mon_list = monsters[highest_cr]
 
-        for mon in monsters[1:]:
-            if mon.cr > highest_cr_mon.cr:
-                highest_cr_mon = mon
+        highest_cr_mon_index = random.get_random_integer(0, len(highest_cr_mon_list) - 1)
+        highest_cr_mon = highest_cr_mon_list[highest_cr_mon_index]
 
         final_list.append(highest_cr_mon)
         cr_budget -= highest_cr_mon.cr
     
-    while cr_budget > 0 and len(final_list) < MONSTER_LIMIT:
-        index = random.get_random_integer(0, monsters_len - 1)
-        mon = monsters[index]
+    while cr_budget > 0 and len(final_list) < MONSTER_LIMIT and len(monsters) > 0:
+        cr_list = getRandomFromDict(monsters)
+        mon = cr_list[random.get_random_integer(0, len(cr_list) - 1)]
         cr_budget -= mon.cr
         final_list.append(mon)
 
+        # remove any monster lists with cr above the budget
+        for key in monsters.keys():
+            if key > cr_budget:
+                monsters.pop(key)
```

Comment:
we should pick which one to remove randomly probably -> fine to do in a later pr

Reply:
But I mean, if the cr is above the remaining budget, they should never be picked anyways, right?
Or do you mean to remove additional lists that could still be picked?

Reply:
if we're removing monsters to stay below the cr budget it shouldn't necessarily be the last one that came in i mean

Reply:
I changed the get_monsters_in_cr_range() to return the monsters partitioned into their cr, so you get a dict with the key 1 and the entry that contains all cr 1 monsters etc
So this step there would remove the list of monsters of a key (=cr) > cr_budget

Reply:
I'm still confused about what exactly you mean, but if that is for a future pr, then we can discuss that in a new issue after this pr is reviewed